### PR TITLE
Fuzzer for simplifying constructors

### PR DESCRIPTION
**Pull request by giltho**
I think I'm happy with the state of the code as-is.
I'm happy with the code as-is. I've basically entirely rewritten the generator, which was the messy part.

The only issue is that we can't merge it because... it's still finding bugs.

- The fuzzer creates expressions and simplifies them, and then checks for equality through z3. Soundness should require that all tests pass, even on corner cases.
- On merge queue, and only when the `bv_solver` folder was modified, the fuzzer is run with 40k tests (8 x 5k)
- On demand, through `!run-bv-fuzzer` or by dispatching the workflow on main, the fuzzer is run with 120k tests (8 x 15k)
- Added a `?force` flag to `eval`, which re-evaluates every node even if children haven't changed.
- Fixed a bunch of unsoundnesses in the simplifications the code found by the fuzzer

### CODE REVIEW

File: soteria/lib/bv_values/svalue.ml [hunk 1]
```diff
@@ -1438,21 +1441,22 @@ and BitVec : BitVec = struct
         zero (size_of v1.node.ty)
     | Binop (Shl, v, { node = { kind = BitVec s1; _ }; _ }), BitVec s2 ->
         let n = size_of v1.node.ty in
-        shl v (mk n Z.(s1 + s2))
+        let total = Z.(s1 + s2) in
+        if Z.(total >= of_int n) then zero n else shl v (mk n total)
```
Comment (N1ark) [inline soteria/lib/bv_values/svalue.ml]: same here there is already a reduction for this so you can just trust `shl`

File: soteria/lib/bv_values/svalue.ml
Comment (N1ark) [inline soteria/lib/bv_values/svalue.ml]: and here too id rather you just remove the `total >= of_int n` branch and add a proper reduction where if the rhs is concrete and `>= n`, it becomes `ashr lhs (mk n (n - 1))`

File: soteria/lib/bv_values/svalue.ml [hunk 2]
```diff
@@ -1992,9 +2009,14 @@ and BitVec : BitVec = struct
   let mul_overflows ~signed v1 v2 =
     match (v1.node.kind, v2.node.kind) with
     | BitVec l, BitVec r -> ovf_check ~signed (size_of v1.node.ty) l r Z.( * )
+    | _ when signed && size_of v1.node.ty == 1 ->
+        (* We need to special-case size one because the other simplifications
+           will mes with that case, and it's pretty easy to simplify. *)
```
Comment (N1ark) [inline soteria/lib/bv_values/svalue.ml]:
```suggestion
        (* We need to special-case size one because the other simplifications
           will mess with that case, and it's pretty easy to simplify. *)
```
